### DISCUSSION TITLE
[Bug]: navigate with unstable_usePrompt, causing `Invalid blocker state transition: unblocked -> proceeding` error.

### DISCUSSION THREAD

**Discussion by woochanleee** [remix-run/react-router#10489] [Q&A] [answered]
### What version of React Router are you using?

6.8.0

### Steps to Reproduce

I created a [CodeSandbox playground](https://codesandbox.io/s/boring-yalow-uhdfvw?file=/src/index.js), which you can access by following the steps below.

1. Click the "toggle" button
2. Click the "submit" button
3. **error occurred**

https://github.com/remix-run/react-router/assets/48552260/67251726-54cf-40ff-a3f8-6edd3368eb5d

### Expected Behavior

No `Invalid blocker state transition: unblocked -> proceeding` error occurs, navigate works well.

### Actual Behavior

`Invalid blocker state transition: unblocked -> proceeding` error occurs, navigate doesn't works.

### ETC

In the linked code sandbox, wrap "navigate" in "setTimeout" seems to work.
But, i would like to know what the underlying solution is, and what caused the problem to occur.

**Comment by kertis15** [comment]
I am using useBlocker, and I have the same issue. Interesting is, that on version 6.11.2 it is working without error, but on version 6.14.0 is not.

**Comment by brophdawg11** [comment]
Does this slightly altered version of `usePrompt` fix your issues?

```js
function usePrompt({ when, message }: { when: boolean; message: string }) {
  let blocker = useBlocker(when);

  React.useEffect(() => {
    if (blocker.state === "blocked") {
      let proceed = window.confirm(message);
      if (proceed) {
        blocker.proceed();
      } else {
        blocker.reset();
      }
    }
  }, [blocker, message]);

  React.useEffect(() => {
    if (blocker.state === "blocked" && !when) {
      blocker.reset();
    }
  }, [blocker, when]);
}
```

**Reply by woochanleee** [reply]
It seems to work fine. But i have a question.

If you compare the snippet to the main branch code, you'll notice that the part below looks different.
So, do you know why setTimeout was used in the main branch code? 
I'm also wondering why don't call clearTimeout in the clean up effect.

```tsx
// your snippet
blocker.proceed();
// main branch
setTimeout(blocker.proceed, 0);
```

**Reply by brophdawg11** [reply]
I'm not sure, I didn't write the initial `usePrompt` implementation and the dev who did has since taken a new role, I'll reach out and see if he recalls.

The one thing I'm not sure of in the provided code sandbox is why you need to manually call `setWhen(false)` in `handleSubmit`?  It feels like you're trying to _manually_ reset the blocker to allow the navigation to proceed - but the idea is that the user chooses to let the navigation to proceed, and upon successful navigation - all blockers are automatically reset internally.

I know this is just a simplification, but I'm curious what the real-world use case is - maybe there is a way to resolve this without the need for `setWhen(false)` in `handleSubmit`?  Currently, the `when` state is completely removed upon navigation to `/exit`, but even if it were in a layout route, you could reset it upon successful navigation if `location.key` changed:

```
const [when, setWhen] = useState(false);
const location = useLocation();
const navigate = useNavigate();

usePrompt({
  when,
  message: "Are you sure you want to leave this page?",
});
useEffect(() => setWhen(false), [location.key]);

const handleSubmit = () => navigate("/exit");
```

**Reply by woochanleee** [reply]
Oh, thank you for your efforts.

Here's my real-world use case.

I have a Form UI inside a Full Modal UI, and the open state of the Modal is passed to the usePrompt when param. This is to get a confim if the user click browser back button.

When the user finishes typing and clicks the submit button, it saves the input and, if successful, changes the modal's open state to false and navigates to `/history` where they can view the submission history.

This is the logic that happens in handleSubmit.

```tsx
const [open, setOpen] = useState(false);

usePrompt({
  when: open,
  message: "Are you sure you want to leave this page?",
});

const handleSubmit = () => {
  fooApiCall({
    onSuccess: () => {
      isOpen(false);
      navigate("/history");
    }
  })
}

return <Modal open={open} onSubmit={handleSubmit} />

```

**Reply by brophdawg11** [reply]
OK - so on submit, after successfully saving the user data, you want to unblock and then navigate without showing `window.confirm`, right?  I think the fix in https://github.com/remix-run/react-router/pull/10687 will still show `window.confirm` even though the data has already been saved, so I'm not sure if that's the right fix for this specific use case or not.

In your case, I do think you want `navigate` to be in a `setTimeout` since you need the `setIsOpen(false)` to unblock the blocker prior to navigating, so it requires react to re-render to update the `usePrompt`.

**Reply by brophdawg11** [reply]
Actually a better approach may be to decouple the modal state/blocker state/submitted state.  I think what you really want here is that we unblock and we navigate away upon successful submission of user data.  Normally this would be done with saving to an action and redirecting, but in your setup I think your navigation is a side effect of the _submission_, not the click:

Here's an updated approach in https://codesandbox.io/s/hopeful-sky-psshcw?file=/src/index.js which allows React to update the blocker based on the `isSubmitted` state prior to navigating.  It's essentially accomplishing the same thing as `setTimeout` but doing so via first-class React primitives and is likely more robust in concurrent mode updates and such.

**Reply by woochanleee** [reply]
> OK - so on submit, after successfully saving the user data, you want to unblock and then navigate without showing window.confirm, right?

Yes !!! Excalty.

> It's essentially accomplishing the same thing as setTimeout but doing so via first-class React primitives and is likely more robust in concurrent mode updates and such.

Oh, thanks for another solution. 

> but in your setup I think your navigation is a side effect of the submission, not the click:

I read [sharing-logic-between-event-handlers](https://react.dev/learn/you-might-not-need-an-effect#sharing-logic-between-event-handlers) in the React docs and wanted to `navigate` in the `onSubmit` handler instead of `useEffect` as described here.

**Reply by louis-young** [reply]
Here's an alternate solution that doesn't introduce a new state value or effect: https://codesandbox.io/s/react-router-useprompt-flushsync-mhq5jg

It uses the `flushSync`API to force React to flush the pending state update immediately. I think both solutions are awesome, but I thought I'd throw this in the mix too 🙂

**Reply by woochanleee** [reply]
I already saw that in the thread below, i think that's a great solution too!

**Comment by louis-young** [comment]
Thanks for raising this issue. I ran into this yesterday.

I had a look at what was going on under the hood and reasoned about this with Matt. I've raised https://github.com/remix-run/react-router/pull/10687 and explained the cause of the bug there 🙂

**Reply by brophdawg11** [reply]
Can you elaborate on your use case @louis-young?  I'm concerned that the fix we're implementing still shows `window.confirm` in the above use case, which doesn't seem like it's expected.  See the alternate approach in https://codesandbox.io/s/hopeful-sky-psshcw?file=/src/index.js and let me know if that also solves your use case?

**Reply by louis-young** [reply]
Your thoughts above are correct as far as I'm aware 👍 

They're updating the state value which affects when the prompt is and isn't shown _and_ navigating from within the same closure. React won't process the state update until the handler function has finished executing, so the confirmation prompt will still be shown after the fix in https://github.com/remix-run/react-router/pull/10687 as React hasn't yet re-rendered the component with the new state value which would ultimately disable the prompt. 

That being said, I still think the fix is worth implementing as I don't think we should enable users to shoot themselves in the foot if you will.

**Reply by louis-young** [reply]
With regards to the alternate approach, I think it works, but, personally, I think the new effect and new state value introduce too much indirection. I think this could be solved in a more concise and idiomatic way using the `flushSync` API to force React to flush the pending state update immediately. Here's an example:

https://codesandbox.io/s/react-router-useprompt-flushsync-mhq5jg

What do you think?

**Reply by brophdawg11** [reply]
I think that's a great solution as well!  I guess what I'm getting at is - do we actually have a "bug" in `usePrompt` or are we just holding it wrong in these cases?

**Reply by brophdawg11** [reply]
(I want to keep the new unit tests regardless - they're great!)

**Reply by louis-young** [reply]
I think it'd be nice to not throw a somewhat obscure exception when somebody uses it in this way. It might not work in the way that they intended it to, but that's just down to React's functional model, closures, and when React processes state updates; people run into issues like this all the time with stale state values within a closure and I don't think it's anything new.

Alternatively, if we're confident that this is the only scenario in which this exception is thrown, then perhaps we throw a more human-friendly exception in this scenario. 

I'm not convinced we should be throwing at all here, but what do you think?

**Reply by brophdawg11** [reply]
I think that's fair - if folks use it in this "incorrect" way (where the intention is to unblock but the navigation still blocks because the unblock + navigation happened synchronously) it would be better to let the block happen instead of throwing.  That at least alerts the user to the underlying issue and they can then fix it with `flushSync`/`isSubmitted`/`setTimeout`/etc.

**Reply by louis-young** [reply]
I think so, that'd get my vote 👍

**Comment by brophdawg11** [comment, marked as answer]
Per the thread above, this was a bit of an add use-case and can be fixed with slight changes to the application logic.  We did merge https://github.com/remix-run/react-router/pull/10687 which will avoid the thrown exception - that should be released in 6.14.3 since we already started the release branch for 6.14.2 before this was merged.

But it's worth noting that this still does not behave as the original issue would have expected.  Unblocking and navigating _in the same event loop_ will still trigger the prompt after the fix.  There are app-land solutions using `flushSync`/`useState`/`setTimeout` in the threads above.

**Reply by brophdawg11** [reply]
We just published version `6.15.0-pre.0` which involves this issue. If you'd like to take it for a test run please try it out and let us know what you think!

**Reply by leonpatry** [reply]
Hi I can confirm that with the latest released version 6.15 we still experienced the issue (in our case using `useBlocker`). The slightly altered version above (resetting blocker when  `blocker.state === "blocked" && !when`) did fix the issue.

**Comment by av-k** [comment]
At version [6.23.1](https://www.npmjs.com/package/react-router-dom/v/6.23.1), I can confirm that this issue also can occurs when, between `useBlocker` has been called and the next call of `blocker.proceed()` the value in the `nextLocation` variable is changed. This can be due to the substitution of any URL query parameters, for example.

In my case the solution was to disable any URL manipulations during working with `useBlocker` logic.
